crates/oxc_allocator/src/pool.rs

```diff
@@ -0,0 +1,68 @@
+use std::mem::ManuallyDrop;
+use std::sync::{Arc, Mutex};
+
+use crate::Allocator;
+
+/// A thread-safe pool for reusing `Allocator` instances to reduce allocation overhead.
+///
+/// Internally uses a `Vec` protected by a `Mutex` to store available allocators.
+#[derive(Default)]
+pub struct AllocatorPool {
+    items: Arc<Mutex<Vec<Allocator>>>,
+}
+
+/// A guard object representing exclusive access to an `Allocator` from the pool.
+///
+/// On drop, the `Allocator` is reset and returned to the pool.
+pub struct AllocatorPoolGuard {
+    item: ManuallyDrop<Allocator>,
+    pool: Arc<Mutex<Vec<Allocator>>>,
+}
+
+impl AllocatorPool {
+    /// Creates a new `AllocatorPool` pre-filled with the given number of default `Allocator` instances.
+    pub fn new_with_capacity(capacity: usize) -> AllocatorPool {
+        let items = std::iter::repeat_with(Allocator::default).take(capacity).collect();
+        AllocatorPool { items: Arc::new(Mutex::new(items)) }
+    }
```

Comment:
I meant more like this:

```rs
pub fn with_capacity(capacity: usize) -> AllocatorPool {
    let items = Vec::with_capacity(capacity);
    AllocatorPool { items: Arc::new(Mutex::new(items)) }
}
```

But yes your way is better! It makes the branch in `items.pop().unwrap_or_default()` completely predictable (unless `plugin-import` is enabled).

I suggest calling this method `new`. "capacity" is a bit misleading, as usually it means "generate the thing with specified capacity, but empty" e.g. `Vec::with_capacity`.